source/dofs/dof_accessor_set.cc

```diff
@@ -41,91 +41,235 @@
+  get_abs(const Number a)
+  {
+    return a;
+  }
+
+  template <typename VectorType>
+  constexpr bool is_dealii_vector =
+    std::is_same<VectorType,
+                 dealii::Vector<typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::BlockVector<typename VectorType::value_type>>::value ||
+    std::is_same<
+      VectorType,
+      dealii::LinearAlgebra::Vector<typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::LinearAlgebra::distributed::Vector<
+                   typename VectorType::value_type>>::value ||
+    std::is_same<VectorType,
+                 dealii::LinearAlgebra::distributed::BlockVector<
+                   typename VectorType::value_type>>::value;
+
+  template <typename T>
+  using set_ghost_state_t =
+    decltype(std::declval<T const>().set_ghost_state(std::declval<bool>()));
+
+  template <typename T>
+  constexpr bool has_set_ghost_state =
+    is_supported_operation<set_ghost_state_t, T>;
+
+  template <typename VectorType,
+            typename std::enable_if<has_set_ghost_state<VectorType>,
+                                    VectorType>::type * = nullptr>
+  void
+  set_ghost_state(VectorType &vector, const bool ghosted)
+  {
+    vector.set_ghost_state(ghosted);
+  }
+
+  template <typename VectorType,
+            typename std::enable_if<!has_set_ghost_state<VectorType>,
+                                    VectorType>::type * = nullptr>
+  void
+  set_ghost_state(VectorType &, const bool)
+  {
+    // serial vector: nothing to do
+  }
+
+  template <int  dim,
+            int  spacedim,
+            bool lda,
+            class OutputVector,
+            typename number>
+  void
+  set_dof_values_with_check(const DoFCellAccessor<dim, spacedim, lda> &cell,
+                            const Vector<number> &local_values,
+                            OutputVector &        values,
+                            const bool            perform_check)
+  {
+    (void)perform_check;
+
+#ifdef DEBUG
+    if (perform_check && is_dealii_vector<OutputVector>)
+      {
+        set_ghost_state(values, true);
+
+        Vector<number> tmp(cell.get_fe().n_dofs_per_cell());
+        cell.get_dof_values(values, tmp);
+
+        for (unsigned int i = 0; i < cell.get_fe().n_dofs_per_cell(); ++i)
+          {
+            // a check consistent with the one in
+            // Utilties::MPI::Partitioner::import_from_ghosted_array_finish()
+            Assert(tmp[i] == number() ||
+                     get_abs(tmp[i] - local_values[i]) <=
+                       get_abs(tmp[i] + local_values[i]) * 100000. *
+                         std::numeric_limits<typename numbers::NumberTraits<
+                           number>::real_type>::epsilon(),
+                   ExcNonMatchingElementsSetDofValuesByInterpolation<number>(
+                     local_values[i], tmp[i]));
+          }
+
+        set_ghost_state(values, false);
```

Comment:
Wouldn't you want to only do this if the state was not ghosted before? In other words, you'd want to reset the state to what it was before.